### PR TITLE
Introducing `UniversalRobustCarbonSolver` after Munhoven (2013)

### PR DESCRIPTION
**Pull request by seamanticscience**
## Purpose 
New robust carbon solver for ocean pH/pCO<sub>2</sub> and tests to compare it to direct cubic and alkalinity correction methods.

### CODE REVIEW

File: src/CarbonSystemSolvers.jl
Comment (navidcy) [inline src/CarbonSystemSolvers.jl:1]: what's happening in this file? I see duplicate code, e.g., twice `include("carbon_chemistry_coefficients.jl")` and more and this results in a bunch of warnings when I do `using ClimaOceanBiogeochemistry`.

Comment (seamanticscience) [reply src/CarbonSystemSolvers.jl:1]: Failed to comment out all the stuff at the end, which was for local testing purposes. Should work now:
```
     Testing Running tests...
Test Summary: | Pass  Total  Time
CarbonSolvers |   28     28  1.4s
warning: essentials.jl:14:0: loop not unrolled: the optimizer was unable to perform the requested transformation; the transformation might be disabled or specified as part of an unsupported transformation ordering
warning: essentials.jl:14:0: loop not unrolled: the optimizer was unable to perform the requested transformation; the transformation might be disabled or specified as part of an unsupported transformation ordering
warning: essentials.jl:14:0: loop not unrolled: the optimizer was unable to perform the requested transformation; the transformation might be disabled or specified as part of an unsupported transformation ordering
warning: essentials.jl:14:0: loop not unrolled: the optimizer was unable to perform the requested transformation; the transformation might be disabled or specified as part of an unsupported transformation ordering
Test Summary:                     | Pass  Total  Time
NutrientsPlanktonBacteriaDetritus |    1      1  1.2s
[ Info: Initializing simulation...
[ Info:     ... simulation initialization complete (49.486 ms)
[ Info: Executing initial time step...
[ Info:     ... initial time step complete (3.824 seconds).
[ Info: Simulation is stopping after running for 4.038 seconds.
[ Info: Model iteration 3 equals or exceeds stop iteration 3.
Test Summary:             | Pass  Total   Time
CarbonAlkalinityNutrients |    3      3  24.0s
     Testing ClimaOceanBiogeochemistry tests passed 
```
Note: `CarbonSystemSolvers` is tested, but not yet used in any examples.

**Comment by seamanticscience** [comment]
For future reference - need to edit `docs/make.jl`, `docs/src/library/internals.md`, and `docs/src/library/public.md` to allow `@autodocs` to hoover up all the docstrings

File: src/CarbonSystemSolvers.jl
Comment (navidcy) [reply src/CarbonSystemSolvers.jl:1]: Do we need those commented code or should we delete them whatsoever?

**Comment by navidcy** [comment]
I merged main (with the latest Oceananigans version) to see if tests still pass or if we need to adjust anything to be up to date with latest Oceananigans syntax. If tests pass I'll approve this.

File: src/CarbonSystemSolvers.jl [hunk 1]
```diff
@@ -556,13 +1090,13 @@ end # end function
 end # module DirectCubicCarbonSolver
 
 # ----------------------------------------------------------------------------------
-
+# ## This should go in the testing suite, eventually.
```
Comment (navidcy) [inline src/CarbonSystemSolvers.jl]: I see it is in the testing suite, right?
so let's remove from here?

**Review by navidcy — APPROVED** [review]
lgtm

it'd be even better if we delete the commented code if it's already included in the tests!

File: src/CarbonSystemSolvers.jl (hunk 1 above)
Comment (seamanticscience) [reply src/CarbonSystemSolvers.jl]: Done

**Comment by glwagner** [comment]
Can we add an example that illustrates the solver in action + illustrates the user interface? For example, generate some 2D synthetic tracer fields, then compute pCO2 and plot it. Ideally illustrating the behavior of the solver in tricky situations and also maybe making a point about the chemistry (ie how pCO2 varies with alkalinity etc)

**Comment by glwagner** [comment]
Possibly the example could pull fields from ECCO and make a global plot?

**Comment by seamanticscience** [comment]
contributes to #15

**Comment by seamanticscience** [comment]
> Can we add an example that illustrates the solver in action + illustrates the user interface? For example, generate some 2D synthetic tracer fields, then compute pCO2 and plot it. Ideally illustrating the behavior of the solver in tricky situations and also maybe making a point about the chemistry (ie how pCO2 varies with alkalinity etc)

See #37

**Pull request merged**